### PR TITLE
Realtime: Expose new API and then refactor a little

### PR DESCRIPTION
**Pull request by gregorylegarec**
## Context
Two weeks ago I talked with @y-lohse about its need of renewing the OAuth token in cozy-realtime. He was arguing that it was not easy to implement a new feature in the existing codebase. I took a look on the code that was first written by myself and damn he was right; the code was a mess, with global variables everywhere, error-prone and subject to race conditions.

I immediately felt guilty and stay obsessed with this lib, I took on me to refactor it a little on my free time. I spent a few night in writing robust end to end tests and then used a class to refactor all the code.

Today @kosssi asked me about realtime and I talked to him about the state of the codebase and this little refacto. That's why I am making this PR now.

## What's changing
The previous API was based only on functions, and was looking like:

```js
import realtime from 'cozy-realtime'


const config = {
  domain: 'cozy.tools:8080',
  token: 'abcdef'
} 

const accountCreationHandler = account => { console.debug('Account has been created', account)}

const accountSubscription = realtime.subscribe(config, 'io.cozy.accounts')
accountSubscription.onCreate(accountCreationHandler)


//...

accountSubscription.unsubscribe()

```
This API is problematic in many ways:

* We must pass a `config` object at every `subscribe()` call
* We can register only one handler for an given event on a given doctype
* We can only unsubscribe to all events related to a doctype

The new API fixes all those problems, and proposes a realtime initialization as suggested in this issue https://github.com/cozy/cozy-libs/issues/10

```js
import { CozyRealtime } from 'cozy-realtime'


const realtime = new CozyRealtime({ url: 'http://cozy.tools:8080', token: 'abcdef')

const accountCreationHandler = account => { console.debug('Account has been created', account)}

realtime.subscribe({ type: 'io.cozy.accounts'}, 'created', accountCreationHandler)

const otherHandler = account => { console.debug(account._id) }
realtime.subscribe({ type: 'io.cozy.accounts'}, 'created', otherHandler)


//...

// Unsubscribe only one handler
realtime.unsubscribe({ type: 'io.cozy.accounts' }, 'created', accountCreationHandler)
```

## New features
This PR introduces also two callbacks to pass to CozyRealtime at initialisation:
* `onError`: Called when a realtime error occurs (can be handler by app)
* `onDisconnect`: Called when CozyRealtime fails to reconnect three times in a row

## Lighter alternative for this PR
If this PR is too big to merge now, as it is not prioritary in our backlogs, we can only merge the first two commits which are introducing the new API based on the legacy code.

## First feedback by @y-lohse 
* Provide graceful disconnection, to call in a `componentDidUnmount` for example
* Handle the case of `this._socketPromise` being `null` in `CozyRealtime._sendSubscribeMessage`

## TODO
- [ ]  Update docs
- [ ] Remove legacy code and release version 3.0
- [ ] Create websocket only when they are one subscription and close websocket on the last unsuscribe

### CODE REVIEW

**Comment by CPatchane** [comment]
We have to add a `BREAKING CHANGE:` label somewhere in a commit description since there will be a new major version for this API change

**Comment by gregorylegarec** [comment]
The only breaking change is a `console.warn()` when using realtime the old way, so I don't know if we have to indicates it now or when the legacy code will be removed.

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+   * Returns an instance of CozyRealtime. Can be used instead of
+   * `new CozyRealtime`.
+   * @static
+   * @param  {Object} options Object containing domain, secure, token and url
+   * parameters
+   * @return {CozyRealtime}         CozyRealtime instance
+   */
+  static init(options) {
+    return new CozyRealtime(options)
+  }
```
Comment (ptbrowne) [inline packages/cozy-realtime/src/CozyRealtime.js]: I think we should remove this. It is idiomatic javascript to write `new`, there is no advantage in having this method, and having several methods to do 1 thing is confusing IMHO.

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+      secure: [isBoolean],
+      token: [isRequired, isString],
+      url: [isRequiredIfNo(['domain']), isURL]
+    }).validate({ domain, secure, token, url })
+
+    this._domain = domain || new URL(url).host
+
+    this._secure = url ? isSecureURL(url) : secure
+    this._token = token
+    this._url = url
```
Comment (ptbrowne) [inline packages/cozy-realtime/src/CozyRealtime.js]: Is it possible not to cache `url` `token` and `secure` and to always go through a cozy-client ?

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+  isString,
+  isURL
+} from './helpers/Validator'
+
+const MAX_RETRIES = 3
+const RETRY_BASE_DELAY = 1000
+
+export class CozyRealtime {
+  /**
+   * Log were subcribe messages sent are recored
```
Comment (ptbrowne) [inline packages/cozy-realtime/src/CozyRealtime.js]:
```suggestion
   * Log were subscribe messages sent are recored
```

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    this._stopListenningUnload()
+
+    if (!event.wasClean) {
+      this._retry(event)
+    }
+  }
+
+  /**
+   * Handle a socket opening, send the authentification message to cozy
+   * stack, and a subribe message for every subscriptions that have been made.
```
Comment (ptbrowne) [inline packages/cozy-realtime/src/CozyRealtime.js]:
```suggestion
   * stack, and a subscribe message for every subscriptions that have been made.
```

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+  static init(options) {
+    return new CozyRealtime(options)
+  }
+
+  /**
+   * Remove the given handler from the list of handlers for given
+   * doctype/document and event.
+   *
+   * @param  {String}  type      Document doctype to subscribe to
+   * @param  {String}  id        Document id to subscribe to
```
Comment (drazik) [inline packages/cozy-realtime/src/CozyRealtime.js]: You can also apply the JSDoc change here

File: packages/cozy-realtime/src/CozyRealtime.js [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+    this._socketPromise = null
+    this._stopListenningUnload()
+
+    if (!event.wasClean) {
+      this._retry(event)
+    }
+  }
+
+  /**
+   * Handle a socket opening, send the authentification message to cozy
```
Comment (drazik) [inline packages/cozy-realtime/src/CozyRealtime.js]: « authentication »

File: packages/cozy-realtime/src/RealtimeSubscriptions.js [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    this._handlers[key] = this._handlers[key] || []
+    if (!this._handlers[key].includes(handler)) {
+      this._handlers[key].push(handler)
+    }
+  }
+
+  /**
+   * Call all the handlers related to the given selector and event name.
+   * @param  {[type]} type      Document doctype
+   * @param  {[type]} id        Document id
```
[the remark this answers is not in the document]
Comment (ptbrowne) [reply packages/cozy-realtime/src/RealtimeSubscriptions.js]: `{[type]}` also

**Comment by nono** [comment]
Ping @gregorylegarec: what is the status of this PR?

**Comment by kosssi** [comment]
@nono I'm working on the next version of cozy-realtime, we put this topic aside but I'm on it now, the goal is to have a connection only if we launched a subscribe

**Comment by gregorylegarec** [comment]
As @kosssi is working on the next steps of this PR, I suggest to close it.
@kosssi just remind to take comment from reviewers into account :)

**Pull request closed without merging**